nntrainer/layers/loss/kld_loss_layer.cpp

```diff
@@ -16,13 +17,61 @@
 #include <vector>
 
 namespace nntrainer {
+static constexpr size_t SINGLE_INOUT_IDX = 0;
+
 KLDLossLayer::KLDLossLayer() {}
 
 KLDLossLayer::~KLDLossLayer() {}
 
-void KLDLossLayer::setProperty(const std::vector<std::string> &values) {}
+void KLDLossLayer::setProperty(const std::vector<std::string> &values) {
+  if (values.size()) {
+    throw std::invalid_argument(
+      "kld loss does not take any properties, but values given");
+  }
+}
+
+void KLDLossLayer::forwarding(RunLayerContext &context, bool training) {
+  // Result = (P * (P / Q).log()).sum()
+  // KL(P ∣∣ Q) whereP denotes the distribution of the observations in datasets
+  // and Q denotes the model output.
+
+  nntrainer::Tensor &predicted = context.getInput(SINGLE_INOUT_IDX);
+  nntrainer::Tensor &output = context.getOutput(SINGLE_INOUT_IDX);
+  if (context.isLabelAvailable(SINGLE_INOUT_IDX)) {
+    nntrainer::Tensor &label = context.getLabel(SINGLE_INOUT_IDX);
+    nntrainer::Tensor temp; // temp output
+    /**
+     * 1. Output = label / predicted
+     * 2. Output = output * label
+     * 3. Output = log(output)
+     * 4. Output = sum(output)
+     */
+    label.divide(predicted, temp);
+    temp.multiply_i(label);
+    temp.apply<float>(logf, temp);
+    output.fill(temp.sum({0, 1, 2, 3}));
+  }
+}
+
+void KLDLossLayer::calcDerivative(RunLayerContext &context) {
+  /**
+   * d/dQ = -P**2/(Q**2) * ln(P/Q)
+   */
+  nntrainer::Tensor &predicted = context.getInput(SINGLE_INOUT_IDX); // Q
```

Comment:
Could you check the derivative formula?
In my opinion,  $KLD(P| Q) = P \log(\frac{P}{Q}) = P \log P - P \log Q$, so $\frac{d}{dQ} KLD(P|Q) = - \frac{P}{Q}$.
(According to the way of reduction, the derivative can be updated as well)

Reply:
We need to revisit our discussion on the formula! Thank you for your insightful feedback.

Reply:
I have completed apply your review.
Thank you for providing additional check with pytorch example.